### PR TITLE
Single source of truth for everything

### PR DESCRIPTION
**Pull request by isiahmeadows**
This stemmed from [this short discussion](https://github.com/fantasyland/fantasy-land/pull/146#issuecomment-244973256). I think it would be a good idea to have a single source of truth for everything: the object itself (instead of an optional `constructor` method as well).

Given other outstanding PRs and issues, I'd expect this would need to wait a bit even if accepted, because it messes with so much. Also, I refactored a few bits of code to not use classes unnecessarily.

### CODE REVIEW

**Comment by isiahmeadows** [comment]
Also, note that this is based on #157, so it carries that commit. [Here's that commit only.](https://github.com/fantasyland/fantasy-land/pull/158/commits/181555e91b11acb2c25e48160693328dbb09bdaf)

**Comment by isiahmeadows** [comment]
(By the way, I'm not that attached to this. If it gets rejected, I'm not that concerned.)

File: README.md [hunk 1]
```diff
@@ -254,27 +252,25 @@ implement the Functor and Foldable specifications.
 1. `t(u.sequence(f.of))` is equivalent to `u.map(t).sequence(g.of)`
 for any `t` such that `t(a).map(f)` is equivalent to `t(a.map(f))` (naturality)
 
-2. `u.map(F.of).sequence(F.of)` is equivalent to `F.of(u)` for any Applicative `F` (identity)
+2. `u.map(f.of).sequence(f.of)` is equivalent to `f.of(u)` for any Applicative `f` (identity)
 
-3. `u.map(x => new Compose(x)).sequence(Compose.of)` is equivalent to
-   `new Compose(u.sequence(F.of).map(v => v.sequence(G.of)))` for `Compose` defined below and any Applicatives `F` and `G` (composition)
+3. `u.map(compose.of).sequence(compose.of)` is equivalent to
```
Comment (rpominov) [inline README.md]: Something wrong here. `new Compose` is intentionally not the same as `Compose.of` see https://github.com/fantasyland/fantasy-land/pull/141

**Comment by rpominov** [comment]
If we have `of` method only on values how do we create a value in the first place? :)

**Comment by rpominov** [comment]
Another ways to improve things in this area:

- Replace "or" with "and" in "on itself or its `constructor`".
- State that in order to be compatible with FL one need to use JS "classes" (either via `function` or `class`) so it would be clear where to look for static `of/empty/etc` — on the class; Require only static `of/empty/etc`.

**Comment by isiahmeadows** [comment]
@rpominov 

> If we have `of` method only on values how do we create a value in the first place? :)

Note that the examples use plain objects for this reason. It's somewhat Lua-like.

> Another ways to improve things in this area:
> 1. Replace "or" with "and" in "on itself or its  constructor ".

I was aiming to remove the latter. (Check out the green parts. `grep . 'constructor'` should bring absolutely nothing.)

> 2. Require only static  of/empty/etc  methods.
>
> Maybe additionally to \#1 or \#2: state that in order to be compatible with FL one need to use JS "classes" (either via `function` or `class`) so it would be clear where to look for static `of/empty/etc` — on the class.


[See this comment by @davidchambers.](https://github.com/fantasyland/fantasy-land/pull/146#issuecomment-245007036) That's why I removed references to constructor methods entirely.

File: README.md (hunk 1 above)
Comment (isiahmeadows) [inline README.md]: I'll fix that when I get home to `compose.create` (the intended replacement method).

**Comment by rpominov** [comment]
> Note that the examples use plain objects for this reason. It's somewhat Lua-like.

I still don't understand how one would use such API. Say someone gave me a type that compatible with FL; how do I create a value of that type? Or maybe FL don't have to support this use case (not sure).

Btw why do we switch to plain objects and `__proto__`? We can remove static methods but keep "classes", right?

> See this comment by @davidchambers. That's why I removed references to constructor methods entirely.

That make sense. Seems like having it on prototype is essential.

Btw, this PR is related to #88 (seems like it wasn't mentioned before).

File: README.md [hunk 2]
```diff
@@ -254,27 +252,25 @@ implement the Functor and Foldable specifications.
...
-Compose.prototype.ap = function(x) {
-  return new Compose(this.c.map(u => y => u.ap(y)).ap(x.c));
-};
-
-Compose.prototype.map = function(f) {
-  return new Compose(this.c.map(y => y.map(f)));
-};
+var compose = {
+  of: function(x) {
+    return {__proto__: compose, c};
```
Comment (SimonRichardson) [inline README.md]: Really?

Comment (isiahmeadows) [inline README.md]: Would you like a different method of setting the prototype? (Keep in mind, I'm not actually that deeply invested in this. I'm merely encouraging exploration.)

**Comment by isiahmeadows** [comment]
@rpominov WRT #88, I think I saw that once, but that's not what drove me to write this. (It is related, though.)

File: README.md (hunk 2 above)
Comment (davidchambers) [inline README.md]: Why not rely on the `constructor` property as we currently do? We could write:

```javascript
//  Identity :: a -> Identity a
var Identity = function Identity(x) {
  return {
    constructor: {'fantasy-land/of': Identity},
    'fantasy-land/chain': function(f) { return f(x); },
    'fantasy-land/map': function(f) { return Identity(f(x)); },
    // ...
    value: x
  };
};
```

The nice thing about this approach is it allows a prototype-based implementation:

```javascript
//  Identity :: a -> Identity a
var Identity = function Identity(x) {
  if (!(this instanceof Identity)) return new Identity(x);
  this.value = x;
};

//  Identity.of :: a -> Identity a
Identity['fantasy-land/of'] = Identity;

//  Identity#chain :: Identity a ~> (a -> Identity b) -> Identity b
Identity.prototype['fantasy-land/chain'] = function(f) { return f(this.value); };

//  Identity#map :: Identity a ~> (a -> b) -> Identity b
Identity.prototype['fantasy-land/map'] = function(f) { return Identity(f(this.value)); };

// ...
```

I don't see a reason to use `__proto__`.

Comment (rpominov) [inline README.md]: > constructor: {'fantasy-land/of': Identity}
> Identity['fantasy-land/of'] = Identity

But didn't we want to remove static method?

Comment (rpominov) [inline README.md]: Anyway if we want to remove static method, I think we should simply _remove static method_ from current implementations, without switching to simple objects or anything.

**Comment by isiahmeadows** [comment]
Given the immediate negative feedback (and the admitted non-idiomatic nature of it), I'm going to close it now. I think this was fully explored.

**Comment by isiahmeadows** [comment]
#157 should still be looked at (it's just a series of bug fixes)

File: README.md [hunk 3]
```diff
@@ -138,11 +138,10 @@ the Semigroup specification.
 empty :: Monoid m => () -> m
 ```
 
-A value which has a Monoid must provide an `empty` method on itself or
-its `constructor` object. The `empty` method takes no arguments:
+A value which has a Monoid must provide an `empty` method. The `empty`
+method takes no arguments:
 
     m.empty()
-    m.constructor.empty()
```
Comment (davidchambers) [inline README.md]: I believe we all agree that *this* change is a good idea. I've opened #162 which makes this change only.

**Pull request closed without merging**